contract-examples/test/utils.ts

```diff
@@ -0,0 +1,123 @@
+ * test("<test_name>", [{ method: "<contract_function_name>", overrides: {}, shouldFail: false, debug: false }])
+ * test("<test_name>", [{ method: "<contract_function_name>", shouldFail: false, debug: false }], {})
+ * ```
+ * All of the above examples are equivalent.
+ * Many contract functions can be called as a part of the same test:
+ * ```js
+ * test("<test_name>", [<test_fn1>, <test_fn2>, <test_fn3>])
+ * ```
+ * Individual test functions can describe their own overrides with the `overrides` property.
+ * If an object is passed in as the third argument to `test`, it will be used as the default overrides for all test
+ * functions.
+ * The following are equivalent:
+ * ```js
+ * test("<test_name>", [{ method: "<contract_function_name>", overrides: { from: "0x123" } }])
+ * test("<test_name>", [{ method: "<contract_function_name>" }], { from: "0x123" })
+ * ```
+ * In the above cases, the `from` override must be a signer.
+ * The `shouldFail` property can be used to indicate that the test function should fail. This should be used sparingly
+ * as it is not possible to match on the failure reason.
+ * Furthermore, the `debug` property can be used to print any thrown errors when attempting to
+ * send a transaction or while waiting for the transaction to be confirmed (the transaction is the smart contract call).
+ * `debug` will also cause any parseable event logs to be printed that start with the `log_` prefix.
+ * `DSTest` contracts have several options for emitting `log_` events.
+ *
+ */
+
+// Below are the types that help define all the different ways to call `test`
+type FnNameOrObject = string | string[] | MethodObject | MethodObject[]
+type MethodObject = { method: string, debug?: boolean, overrides?: any, shouldFail?: boolean }
+
+// This type is after all default values have been applied
+type MethodWithDebugAndOverrides = MethodObject & { debug: boolean, overrides: any, shouldFail: boolean }
+
+// `test` is used very similarly to `it` from Mocha
+export const test = (name, fnNameOrObject, overrides = {}) => it(name, testFn(fnNameOrObject, overrides))
+// `test.only` is used very similarly to `it.only` from Mocha, it will isolate all tests marked with `test.only`
+test.only = (name, fnNameOrObject, overrides = {}) => it.only(name, testFn(fnNameOrObject, overrides))
+// `test.debug` is used to apply `debug: true` to all DSTest contract method calls in the test
+test.debug = (name, fnNameOrObject, overrides = {}) => it.only(name, testFn(fnNameOrObject, overrides, true))
+// `test.skip` is used very similarly to `it.skip` from Mocha, it will skip all tests marked with `test.skip`
+test.skip = (name, fnNameOrObject, overrides = {}) => it.skip(name, testFn(fnNameOrObject, overrides))
+
+// `testFn` is a higher-order function. It returns a function that can be used as the test function for `it`
+const testFn = (fnNameOrObject: FnNameOrObject, overrides = {}, debug = false) => {
```

Comment:
It's not clear `testFn` by the name created a higher order function

Reply:
I mean, I'll change the name as this comment is a little overly pedantic. On line 55, it literally says `testFn` is a higher-order function.